File: test-apps/display-test-app/ios/imodeljs-test-app/imodeljs-test-app/DtaServiceAuthorizationClient.swift
/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import Foundation

/// A struct to hold the settings used by DtaServiceAuthSettings
struct DtaServiceAuthSettings {
    var clientId: String
    var clientSecret: String
    var scope: String
    var authority: String
}

typealias DtaServiceAuthorizationClientCallback = (Error?) -> ()

class DtaServiceAuthorizationClient: NSObject, DtaAuthorizationClient {
    /// Instance for `onAccessTokenChanged` property from the `AuthorizationClient` protocol.
    public var onAccessTokenChanged: AccessTokenChangedCallback?
    /// Instance for `errorDomain` property from the `DtaAuthorizationClient` protocol.
    public let errorDomain = "com.bentley.display-test-app"
    var issuerURL = URL(string: "https://ims.bentley.com")!
    var authSettings: DtaServiceAuthSettings
    var userInfo: NSDictionary?
    var accessToken: String?
    var expirationDate: Date?

    init?(configData: JSON) {
        let clientId = configData["IMJS_OIDC_CLIENT_ID"] as? String ?? ""
        let clientSecret = configData["IMJS_OIDC_CLIENT_SECRET"] as? String ?? ""
        let scope = configData["IMJS_OIDC_SCOPE"] as? String ?? ""
        let authority = configData["IMJS_SERVICE_AUTHORITY"] as? String ?? "https://ims.bentley.com"
        authSettings = DtaServiceAuthSettings(clientId: clientId, clientSecret: clientSecret, scope: scope, authority: authority)
        super.init()
        guard (try? checkSettings()) != nil else {
            return nil
        }
        if var prefix = configData["IMJS_URL_PREFIX"] as? String,
           let authorityURL = URL(string: authority),
           var authorityURLComponents = URLComponents(url: authorityURL, resolvingAgainstBaseURL: false) {
            prefix = prefix == "dev-" ? "qa-" : prefix
            authorityURLComponents.host = prefix + authorityURLComponents.host!
            if authorityURLComponents.path.hasSuffix("/") {
                authorityURLComponents.path.removeLast()
            }
            if let url = authorityURLComponents.url {
                self.issuerURL = url
            }
        }
    }

    func checkSettings() throws {
        if authSettings.clientId.count == 0 {
            throw error(reason: "DtaServiceAuthSettings: initialize() was called with invalid or empty clientId")
        }
        if authSettings.clientSecret.count == 0 {
            throw error(reason: "DtaServiceAuthSettings: initialize() was called with invalid or empty clientSecret")
        }
        if authSettings.scope.count == 0 {
            throw error(reason: "DtaServiceAuthSettings: initialize() was called with invalid or empty scope")
        }
        let scope = authSettings.scope
        if scope.contains("openid") || scope.contains("email") || scope.contains("profile") || scope.contains("organization") {
            throw error(reason: "DtaServiceAuthSettings: Scopes for a service cannot include 'openid email profile organization'")
        }
    }

    func generateAccessToken(_ completion: @escaping DtaServiceAuthorizationClientCallback) {
        let session = URLSession(configuration: URLSessionConfiguration.default)
        let requestURL = issuerURL.appendingPathComponent("connect").appendingPathComponent("token")
        var request = URLRequest(url: requestURL)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpMethod = "POST"
        let bodyString = "grant_type=client_credentials&scope=\(authSettings.scope.encodedForURLQuery()!)"
        let bodyData = Data(bodyString.utf8)
        request.setValue("\(bodyData.count)", forHTTPHeaderField: "Content-Length")
        request.httpBody = bodyData
        let encoded = "\(authSettings.clientId.encodedForURLQuery()!):\(authSettings.clientSecret.encodedForURLQuery()!)"
        request.setValue("Basic \(encoded.toBase64())", forHTTPHeaderField: "Authorization")
        let task = session.dataTask(with: request) { (data, response, error) in
            if let error = error {
                completion(error)
                return
            }
            guard let response = response as? HTTPURLResponse else {
                completion(self.error(reason: "Invalid response."))
                return
            }
            guard response.statusCode == 200 else {
                completion(self.error(reason: "Status code: \(response.statusCode)"))
                return
            }
            guard let data = data, let json = try? JSONSerialization.jsonObject(with: data) as? JSON else {
                completion(self.error(reason: "Invalid response body."))
                return
            }

            self.accessToken = (json["access_token"] as? String).map { "Bearer \($0)" }
            self.expirationDate = (json["expires_in"] as? Double).map { Date(timeIntervalSinceNow: $0) }
            completion(nil)
        }
        task.resume()
    }

    // MARK: - AuthorizationClient Protocol implementation
    public func getAccessToken(_ completion: @escaping GetAccessTokenCallback) {
        if accessToken != nil, expirationDate != nil {
            completion(accessToken, expirationDate, nil)
        }
        generateAccessToken() { error in
            completion(self.accessToken, self.expirationDate, error)
        }
    }
}

File: test-apps/display-test-app/android/imodeljs-test-app/app/src/main/java/com/bentley/imodeljs_test_app/DtaServiceAuthorizationClient.kt
package com.bentley.imodeljs_test_app

import android.net.Uri
import android.util.Base64
import com.bentley.itwin.AuthTokenCompletionAction
import com.bentley.itwin.AuthorizationClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.DataOutputStream
import java.net.URL
import java.util.*
import javax.net.ssl.HttpsURLConnection

class DtaServiceAuthorizationClient(env: JSONObject): AuthorizationClient() {
    data class AuthSettings(val clientId: String, val clientSecret: String, val scope: String, val authority: String)
    private val authSettings = parseEnv(env)
    private var accessToken: String? = null
    private var expirationDate: String? = null

    companion object {
        fun create(env: JSONObject): DtaServiceAuthorizationClient? {
            return try {
                DtaServiceAuthorizationClient(env)
            } catch (ex: Exception) {
                null
            }
        }
    }
    private fun parseEnv(env: JSONObject): AuthSettings {
        val clientId = env.optStringNotEmpty("IMJS_OIDC_CLIENT_ID")
        val clientSecret = env.optStringNotEmpty("IMJS_OIDC_CLIENT_SECRET")
        val scope = env.optStringNotEmpty("IMJS_OIDC_SCOPE")
        val authority = env.optStringNotEmpty("IMJS_SERVICE_AUTHORITY") ?: "https://ims.bentley.com"
        if (clientId == null || clientSecret == null || scope == null) {
            throw Exception("env does not contain necessary config data.")
        }
        val scopes = scope.split(" ")
        if (scopes.contains("openid") || scopes.contains("email") || scopes.contains("profile") || scopes.contains("organization")) {
            throw Exception("DtaServiceAuthSettings: Scopes for a service cannot include 'openid email profile organization'")
        }
        return AuthSettings(clientId, clientSecret, scope, authority)
    }

    private suspend fun generateAccessToken() {
        return withContext(Dispatchers.IO) {
            val authURL = URL("${authSettings.authority}/connect/token")
            with (authURL.openConnection() as HttpsURLConnection) {
                useCaches = false
                doOutput = true
                requestMethod = "POST"
                val encodedScope = Uri.encode(authSettings.scope)
                val body = "grant_type=client_credentials&scope=${encodedScope}"
                val bodyData = body.toByteArray()
                setRequestProperty("Accept", "application/json")
                setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
                setRequestProperty("charset", "utf-8")
                setRequestProperty("Content-Length", "${bodyData.size}")
                val encodedClientId = Uri.encode(authSettings.clientId)
                val encodedClientSecret = Uri.encode(authSettings.clientSecret)
                val encoded = "${encodedClientId}:${encodedClientSecret}"
                setRequestProperty("Authorization", "Basic ${Base64.encodeToString(encoded.toByteArray(), Base64.NO_WRAP)}")
                val os = DataOutputStream(outputStream)
                os.write(bodyData)
                os.flush()
                if (responseCode != 200) {
                    val message = "Wrong response code fetching token: $responseCode"
                    throw Exception(message)
                }
                val response = JSONObject(inputStream.bufferedReader().use(BufferedReader::readText))
                val responseValues = listOfNotNull(response.optStringNotEmpty("access_token"), response.optStringNotEmpty("expires_in"))
                if (responseValues.size != 2) {
                    throw Exception("Did not get token and expiration date in response.")
                }
                accessToken = "Bearer ${responseValues[0]}"
                val expiresIn = responseValues[1].toLong()
                expirationDate = Date(Date().time + expiresIn * 1000).toString()
            }
        }
    }

    private fun returnAccessToken(tokenAction: AuthTokenCompletionAction): Boolean {
        return let2(accessToken, expirationDate) {
            tokenAction.resolve(it.first, it.second)
            true
        } == true
    }

    override fun getAccessToken(tokenAction: AuthTokenCompletionAction) {
        if (returnAccessToken(tokenAction)) return
        MainScope().launch {
            try {
                generateAccessToken()
                if (!returnAccessToken(tokenAction)) {
                    tokenAction.error("Could not fetch access token")
                }
            } catch (ex: Exception) {
                tokenAction.error(ex.toString())
            }
        }
    }
}

inline fun <T1: Any, T2: Any, R: Any> let2(p1: T1?, p2: T2?, block: (Pair<T1, T2>)->R?): R? {
    return if (p1 != null && p2 != null) block(Pair(p1, p2)) else null
}
